File: 3600-agents/Bobby_4/zobrist.py
# zobrist.py

from __future__ import annotations

import random
from typing import List, Tuple, Set
from game import board as board_mod  # type: ignore

# Feature indices (POV-relative)
Z_ME_EGG      = 0
Z_OPP_EGG     = 1
Z_ME_TURD     = 2
Z_OPP_TURD    = 3
Z_ME_CHICKEN  = 4
Z_OPP_CHICKEN = 5
Z_KNOWN_TRAP  = 6

Z_NUM_FEATURES = 7

# Global tables
_ZOBRIST_TABLE: List[List[List[int]]] | None = None
_ZOBRIST_SIDE_TO_MOVE: int | None = None


def init_zobrist(dim: int = 8, seed: int = 1234567) -> None:
    """
    Initialize Zobrist random bitstrings.

    Must be called once at startup (before using zobrist_hash).
    Using a fixed seed makes behavior deterministic for debugging.
    """
    global _ZOBRIST_TABLE, _ZOBRIST_SIDE_TO_MOVE

    rnd = random.Random(seed)
    _ZOBRIST_TABLE = [
        [
            [rnd.getrandbits(64) for _ in range(Z_NUM_FEATURES)]
            for _ in range(dim)
        ]
        for _ in range(dim)
    ]
    _ZOBRIST_SIDE_TO_MOVE = rnd.getrandbits(64)


def zobrist_hash(
    cur_board: board_mod.Board,
    known_traps: Set[Tuple[int, int]],
) -> int:
    """
    Optimized Sparse Zobrist Hash.
    Iterates over piece sets instead of the 8x8 grid.
    Speedup: ~20x faster than the dense loop.
    """
    global _ZOBRIST_TABLE, _ZOBRIST_SIDE_TO_MOVE

    if _ZOBRIST_TABLE is None or _ZOBRIST_SIDE_TO_MOVE is None:
        raise RuntimeError("Zobrist not initialized. Call init_zobrist() first.")

    table = _ZOBRIST_TABLE
    h = 0

    # 1. Chickens (Always 2)
    mx, my = cur_board.chicken_player.get_location()
    h ^= table[mx][my][Z_ME_CHICKEN]

    ox, oy = cur_board.chicken_enemy.get_location()
    h ^= table[ox][oy][Z_OPP_CHICKEN]

    # 2. My Eggs (Sparse Set Iteration)
    for x, y in cur_board.eggs_player:
        h ^= table[x][y][Z_ME_EGG]

    # 3. Opp Eggs
    for x, y in cur_board.eggs_enemy:
        h ^= table[x][y][Z_OPP_EGG]

    # 4. My Turds
    for x, y in cur_board.turds_player:
        h ^= table[x][y][Z_ME_TURD]

    # 5. Opp Turds
    for x, y in cur_board.turds_enemy:
        h ^= table[x][y][Z_OPP_TURD]

    # 6. Known Traps
    for x, y in known_traps:
        h ^= table[x][y][Z_KNOWN_TRAP]

    # 7. Side to move (Constant XOR)
    h ^= _ZOBRIST_SIDE_TO_MOVE

    return h

class TTEntry:
    __slots__ = ("value", "depth", "flag", "best_move", "gen")

    def __init__(self, value: float, depth: int, flag: str, best_move, gen: int):
        self.value = value
        self.depth = depth
        self.flag = flag     # "EXACT" / "LOWER" / "UPPER"
        self.best_move = best_move
        self.gen = gen

File: 3600-agents/CPPHikaru_3/zobrist.cpp
#include "zobrist.h"
#include <algorithm>
#include <cstdlib>

uint64_t ZobristHash::table[MAP_SIZE][MAP_SIZE][Z_NUM_FEATURES];
uint64_t ZobristHash::side_to_move = 0;
bool ZobristHash::initialized = false;

uint64_t ZobristHash::random_uint64(uint64_t& seed) {
    seed ^= seed << 13;
    seed ^= seed >> 7;
    seed ^= seed << 17;
    return seed;
}

void ZobristHash::init(int dim, uint64_t seed) {
    if (initialized) return;
    
    uint64_t rng_state = seed;
    
    for (int x = 0; x < MAP_SIZE; ++x) {
        for (int y = 0; y < MAP_SIZE; ++y) {
            for (int f = 0; f < Z_NUM_FEATURES; ++f) {
                table[x][y][f] = random_uint64(rng_state);
            }
        }
    }
    
    side_to_move = random_uint64(rng_state);
    initialized = true;
}

uint64_t ZobristHash::hash(const GameState& state, Bitboard known_traps) {
    if (!initialized) {
        init(MAP_SIZE);
    }
    
    uint64_t h = 0;
    
    // Hash chickens
    Position my_pos = state.chicken_player_pos;
    h ^= table[my_pos.x][my_pos.y][Z_ME_CHICKEN];
    
    Position opp_pos = state.chicken_enemy_pos;
    h ^= table[opp_pos.x][opp_pos.y][Z_OPP_CHICKEN];
    
    // Hash eggs (iterate through set bits)
    Bitboard eggs = state.eggs_player;
    while (eggs) {
        int bit = BitboardOps::bsf(eggs);
        Position pos = BitboardOps::bit_to_pos(bit);
        h ^= table[pos.x][pos.y][Z_ME_EGG];
        eggs &= eggs - 1; // Clear least significant bit
    }
    
    eggs = state.eggs_enemy;
    while (eggs) {
        int bit = BitboardOps::bsf(eggs);
        Position pos = BitboardOps::bit_to_pos(bit);
        h ^= table[pos.x][pos.y][Z_OPP_EGG];
        eggs &= eggs - 1;
    }
    
    // Hash turds
    Bitboard turds = state.turds_player;
    while (turds) {
        int bit = BitboardOps::bsf(turds);
        Position pos = BitboardOps::bit_to_pos(bit);
        h ^= table[pos.x][pos.y][Z_ME_TURD];
        turds &= turds - 1;
    }
    
    turds = state.turds_enemy;
    while (turds) {
        int bit = BitboardOps::bsf(turds);
        Position pos = BitboardOps::bit_to_pos(bit);
        h ^= table[pos.x][pos.y][Z_OPP_TURD];
        turds &= turds - 1;
    }
    
    // Hash known traps
    while (known_traps) {
        int bit = BitboardOps::bsf(known_traps);
        Position pos = BitboardOps::bit_to_pos(bit);
        h ^= table[pos.x][pos.y][Z_KNOWN_TRAP];
        known_traps &= known_traps - 1;
    }
    
    // Side to move
    h ^= side_to_move;
    
    return h;
}

